### PR TITLE
Analyze properties with conditional attributes in patterns

### PR DESCRIPTION
**Pull request by jcouv**
`LearnFromDecisionDag` is the method that analyzes the pattern DAG and figures out the nullable state for each label in the DAG.

Three main parts of the change:
1. Previously, that method assumed that we're starting with an unconditional state. Now, it can handle the split state resulting (such as that from a property marked with `MemberNotNullWhen`).
2. The DAG for `{ IsOk : true }` is composed of an evaluation step (that evaluates the property) and a test step. Now, it is possible for the evaluation step to result in a split state. So the data structure that keeps track of the worst case state as input for each DAG node now has to store states that may be conditional.
3. We'd like the analysis of `(IsOk, ...) is (true, ...)` to behave like that of `IsOk is true` despite temps and tuples being involved. So we re-use the slots of the element expressions when possible. Note this only works for top-level tuple, not for nested tuple.

Filed issue https://github.com/dotnet/roslyn/issues/50980 for equality on tuples. We'd like the analysis of `(IsOk, ...) == (true, ...)` to behave like that of `IsOk == true`.

Fixes https://github.com/dotnet/roslyn/issues/49750
Fixes https://github.com/dotnet/roslyn/issues/51020 (patterns on tuple inputs).

### CODE REVIEW

**Comment by jcouv** [comment]
>                 initialState[originalInputSlot] = expressionType.State;

:memo: we use the indexer on `PossiblyConditionalState` here.

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:336 in 6f09fcc. [](commit_id = 6f09fcc3922bf43667276948ef4c45831ed57df7, deletion_comment = False)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs [hunk 1]
```diff
@@ -246,13 +247,66 @@ protected override void VisitSwitchSection(BoundSwitchSection node, bool isLastS
...
+                    result.State = default;
+                    result.StateWhenTrue = StateWhenTrue.Clone();
+                    result.StateWhenFalse = StateWhenFalse.Clone();
+                }
+                else
+                {
+                    result.IsConditionalState = false;
+                    result.State = State.Clone();
+                    result.StateWhenTrue = default;
+                    result.StateWhenFalse = default;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:296]: `return new PossiblyConditionalState(State);`

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs [hunk 2]
```diff
@@ -246,13 +247,66 @@ protected override void VisitSwitchSection(BoundSwitchSection node, bool isLastS
...
+                    StateWhenFalse = nullableWalker.StateWhenFalse.Clone();
+                    IsConditionalState = true;
+                    State = default;
+                }
+                else
+                {
+                    StateWhenTrue = StateWhenFalse = default;
+                    IsConditionalState = false;
+                    State = nullableWalker.State.Clone();
+                }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:278]: These cases are duplicated in the constructor above and `Clone()` below. Consider adding another constructor, `PossiblyConditionalState(LocalState whenTrue, LocalState whenFalse)`, and change this method to a `static` factory method that calls one of the two constructors depending on `IsConditionalState`.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs [hunk 3]
```diff
@@ -246,13 +247,66 @@ protected override void VisitSwitchSection(BoundSwitchSection node, bool isLastS
...
+
+            public PossiblyConditionalState Clone()
+            {
+                PossiblyConditionalState result;
+                if (IsConditionalState)
+                {
+                    result.IsConditionalState = true;
+                    result.State = default;
+                    result.StateWhenTrue = StateWhenTrue.Clone();
+                    result.StateWhenFalse = StateWhenFalse.Clone();
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:289]: `return new PossiblyConditionalState(StateWhenTrue, StateWhenFalse);`

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs [hunk 4]
```diff
@@ -263,15 +317,24 @@ protected override void VisitSwitchSection(BoundSwitchSection node, bool isLastS
...
-                initialState[originalInputSlot] = expressionType.State;
             }
+            Debug.Assert(originalInputSlot > 0);
+
+            // If the input of the switch (or is-pattern expression) is a tuple literal, we reuse the slots of
+            // those expressions (when possible), pretending that we are not copying them into a temporary ValueTuple instance
+            // to evaluate the patterns.  In this way we infer non-nullability of the original element's parts.
+            // We do not extend such courtesy to nested tuple literals.
+            var originalInputElementSlots = expression is BoundTupleExpression tuple
+                ? tuple.Arguments.SelectAsArray(a => MakeSlot(a))
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:328]: >SelectAsArray(a => MakeSlot(a)) [](start = 34, length = 31)

Consider passing in the walker: `SelectAsArray(static (a, w) => w.MakeSlot(a), this)`

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs [hunk 5]
```diff
@@ -341,11 +412,40 @@ protected override void VisitSwitchSection(BoundSwitchSection node, bool isLastS
...
+                                        if (outputSlot <= 0)
+                                        {
+                                            outputSlot = GetOrCreateSlot(field, inputSlot, forceSlotEvenIfEmpty: true);
+
+                                            if (originalTupleElement is not null)
+                                            {
+                                                // The expression in the tuple could not be assigned a slot (for example, `a?.b`),
+                                                // so we had to create a slot for the tuple element instead.
+                                                // We'll remember that so that we can apply any learnings to the expression.
+                                                if (!originalInputMap.ContainsKey(outputSlot))
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:433]: >outputSlot [](start = 82, length = 10)

`GetOrCreateSlot()` may return `-1`.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs [hunk 6]
```diff
@@ -246,13 +247,66 @@ protected override void VisitSwitchSection(BoundSwitchSection node, bool isLastS
...
+        }
+
         private PooledDictionary<LabelSymbol, (LocalState state, bool believedReachable)>
             LearnFromDecisionDag(
             SyntaxNode node,
             BoundDecisionDag decisionDag,
             BoundExpression expression,
             TypeWithState expressionType,
-            ref LocalState initialState)
+            ref PossiblyConditionalState initialState)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:309]: >PossiblyConditionalState [](start = 16, length = 24)

Is this change necessary? It looks like all callers simply convert a `LocalState` to a `PossiblyConditionalState`. If so, consider converting inside the method.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 7]
```diff
@@ -24605,6 +24605,1499 @@ void Init()
...
+{
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk { get; }
+    public string? Value { get; }
+
+    public string M(object? o)
+    {
+        return (this, o) switch
+        {
+            ({ IsOk: true }, null) => M2(Value), // incorrect warning
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:24830]: >incorrect warning [](start = 52, length = 17)

What does this comment refer to?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 8]
```diff
@@ -24605,6 +24605,1499 @@ void Init()
...
+            o2.ToString(); // 2
+        }
+    }
+
+    void M4(object? o, object o2)
+    {
+        if ((true, (o, o2)) is (true, (not null, null)))
+        {
+            o.ToString(); // 3
+            o2.ToString();
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:24949]: Shouldn't the warning be reported for `o2` rather than `o`?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 9]
```diff
@@ -24605,6 +24605,1499 @@ void Init()
...
+
+            c.VerifyDiagnostics(
+                // (21,24): warning CS8603: Possible null reference return.
+                //                 return Value; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "Value").WithLocation(21, 24)
+                );
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_RecursivePattern_WithNullPattern()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:25246]: >MemberNotNullWhenTrue_RecursivePattern_WithNullPattern [](start = 20, length = 54)

Does this test cover a distinct scenario from the previous test?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 10]
```diff
@@ -24605,6 +24605,1499 @@ void Init()
...
+                //                 return value; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "value").WithLocation(18, 24),
+                // (20,17): warning CS0162: Unreachable code detected
+                //                 return Value; // 2
+                Diagnostic(ErrorCode.WRN_UnreachableCode, "return").WithLocation(20, 17)
+                );
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_RecursivePattern_SideEffectOnStruct()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:25495]: >MemberNotNullWhenTrue_RecursivePattern_SideEffectOnStruct [](start = 20, length = 57)

Is this test needed? And where is the side-effect?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 11]
```diff
@@ -24605,6 +24605,1499 @@ void Init()
...
+                //             return Value; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "Value").WithLocation(21, 20),
+                // (24,16): warning CS8603: Possible null reference return.
+                //         return Value; // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "Value").WithLocation(24, 16)
+                );
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_BinaryPatternTest_SideEffectOnStruct()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:25703]: >MemberNotNullWhenTrue_BinaryPatternTest_SideEffectOnStruct [](start = 20, length = 58)

Is this test needed? (Does pattern matching differentiate members that have side-effects?) And where is the side-effect?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 12]
```diff
@@ -24605,6 +24605,1499 @@ void Init()
...
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk { get; }
+    public string? Value { get; }
+
+    public string M()
+    {
+        return IsOk switch
+        {
+            true => Value,
+            _ => throw null!,
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:25953]: >_ => throw null!, [](start = 12, length = 17)

Consider testing `_ => Value` instead.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 13]
```diff
@@ -24605,6 +24605,1499 @@ void Init()
...
+        };
+    }
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_SwitchExpressionOnProperty_WhenFalseBranch()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:25990]: >MemberNotNullWhenTrue_SwitchExpressionOnProperty_WhenFalseBranch [](start = 20, length = 64)

Consider merging into `MemberNotNullWhenTrue_SwitchExpressionOnProperty`.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 14]
```diff
@@ -24605,6 +24605,1499 @@ void Init()
...
+
+public class C
+{
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk { get; }
+    public string? Value { get; }
+
+    public string M()
+    {
+        return IsOk ? Value : throw null!;
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:26031]: >throw null! [](start = 30, length = 11)

Consider testing `Value` instead, and merge with test below.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs (hunk 6 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:309]: Yes, we need pass in a conditional state as the starting point of DAG analysis. That's how  `IsOk is true` and `IsOk switch { true => ... }` and `switch (IsOk) { ... }` work.

All three callers do `new PossiblyConditionalState(this)` to capture the state after the receiver is evaluated, so we could factor that instance creation here, but I think it's better to let the callers do it, to make it explicit what information is passed to this method.

---
In reply to: [573089929](https://github.com/dotnet/roslyn/pull/51001#discussion_r573089929) [](ancestors = 573089929)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 8 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:24949]: We're dealing with a nested tuple here, which makes us lose track. So the test `o is not null` does not update the state of `o` (that's why we still get a warning on `o.ToString()`). Similarly, the pure null test on `o2` doesn't actually affect `o2`, but instead some temp.
I'll add a comment to clarify that the behavior is not what we'd prefer.
Let me know if you think I should file a follow-up issue. I didn't prioritize this scenario because I think nested tuples are not very useful in pattern-matching...

---
In reply to: [573098133](https://github.com/dotnet/roslyn/pull/51001#discussion_r573098133) [](ancestors = 573098133)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 9 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:25246]: The above test follows the `IsOk` check with a boolean test, whereas this one follows it with a null test. I was trying to cover difference code paths, but I don't think it ended mattering. I could remove.

---
In reply to: [573105354](https://github.com/dotnet/roslyn/pull/51001#discussion_r573105354) [](ancestors = 573105354)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 10 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:25495]: This test illustrates that we're not testing a copy of the struct, but the original. Although this doesn't reflect the lowered code, we can do that because analysis of patterns doesn't care about properly understanding side-effects.

---
In reply to: [573108996](https://github.com/dotnet/roslyn/pull/51001#discussion_r573108996) [](ancestors = 573108996)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 11 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:25703]: Same comment as `MemberNotNullWhenTrue_RecursivePattern_SideEffectOnStruct`, except that we're dealing with an `is` rather than a `switch`.

---
In reply to: [573113257](https://github.com/dotnet/roslyn/pull/51001#discussion_r573113257) [](ancestors = 573113257)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 14 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:26031]: The diagnostic produced by a merged test is rather bad to pinpoint what I'm trying to show with these tests:
```
                    // (13,16): warning CS8603: Possible null reference return.
                    //         return IsOk 
                    Diagnostic(ErrorCode.WRN_NullReferenceReturn, @"IsOk 
                ? Value 
                : Value").WithLocation(13, 16)
```

---
In reply to: [573116876](https://github.com/dotnet/roslyn/pull/51001#discussion_r573116876) [](ancestors = 573116876)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 13 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:25990]: Same problem. The diagnostic is on the return, but I want to show which branch the problem comes from.

---
In reply to: [573116282](https://github.com/dotnet/roslyn/pull/51001#discussion_r573116282) [](ancestors = 573116282)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 12 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:25953]: This results in a diagnostic on the `return`, which doesn't tell us which branch caused the problem:
```
                 // (13,16): warning CS8603: Possible null reference return.
                    //         return IsOk switch
                    Diagnostic(ErrorCode.WRN_NullReferenceReturn, @"IsOk switch
            {
                true => Value,
                _ => Value, // 1
            }").WithLocation(13, 16)
```

---
In reply to: [573115096](https://github.com/dotnet/roslyn/pull/51001#discussion_r573115096) [](ancestors = 573115096)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs [hunk 15]
```diff
@@ -246,13 +247,55 @@ protected override void VisitSwitchSection(BoundSwitchSection node, bool isLastS
...
+            public PossiblyConditionalState(LocalState state)
+            {
+                StateWhenTrue = StateWhenFalse = default;
+                IsConditionalState = false;
+                State = state.Clone();
+            }
+
+            public static PossiblyConditionalState Create(NullableWalker nullableWalker)
+            {
+                if (nullableWalker.IsConditionalState)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:274]: >if [](start = 16, length = 2)

Minor: Consider using `?:` rather than `if`, to match `Clone()` below.

**Review by cston — APPROVED** [review]

**Comment by jcouv** [comment]
@RikkiGibson @dotnet/roslyn-compiler for a second review. Thanks

**Comment by RikkiGibson** [comment]
Taking a look

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs [hunk 16]
```diff
@@ -246,13 +247,55 @@ protected override void VisitSwitchSection(BoundSwitchSection node, bool isLastS
             VisitStatementList(node);
         }
 
+        private struct PossiblyConditionalState
+        {
+            public LocalState State;
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:252]: randomly, I wonder if this field can be collapsed into `StateWhenTrue`, and accessed via properties which assert on the value of `IsConditionalState`. No need to try and change things around to do it, just a thought that entered my mind.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs [hunk 17]
```diff
@@ -341,11 +401,40 @@ protected override void VisitSwitchSection(BoundSwitchSection node, bool isLastS
                                         var field = (FieldSymbol)AsMemberOfType(inputType, e.Field);
                                         var type = field.TypeWithAnnotations;
                                         var output = new BoundDagTemp(e.Syntax, type.Type, e);
-                                        int outputSlot = GetOrCreateSlot(field, inputSlot, forceSlotEvenIfEmpty: true);
+                                        int outputSlot = -1;
+                                        var originalTupleElement = e.Input.IsOriginalInput && !originalInputElementSlots.IsDefault
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:405]: Should the declaration of `originalInputElementSlots` be moved closer to the usage?

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:405]: oh, I think that its current place is the only way to avoid recomputing it. Disregard this :)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs [hunk 18]
```diff
@@ -410,7 +507,7 @@ protected override void VisitSwitchSection(BoundSwitchSection node, bool isLastS
                                         learnFromNonNullTest(inputSlot, ref this.StateWhenTrue);
                                     }
                                     gotoNode(p.WhenTrue, this.StateWhenTrue, nodeBelievedReachable);
-                                    gotoNode(p.WhenFalse, this.StateWhenFalse, nodeBelievedReachable & inputState.MayBeNull());
+                                    gotoNode(p.WhenFalse, this.StateWhenFalse, nodeBelievedReachable & inputMaybeNull);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:510]: It looks like this introduces the effect where, if an input has not-null flow state, we assume that dag nodes which rely on the input actually being `null` are unreachable. But in general when the user null tests an input, we sort of do the opposite and say: oh, now we know that it actually could be null. Which scenarios are affected by this?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs [hunk 19]
```diff
@@ -363,6 +452,13 @@ protected override void VisitSwitchSection(BoundSwitchSection node, bool isLastS
                                         }
                                         Debug.Assert(outputSlot > 0);
                                         addToTempMap(output, outputSlot, type.Type);
+
+                                        if (property.GetMethod is not null)
+                                        {
+                                            // A property evaluation splits the state if MemberNotNullWhen is used
+                                            ApplyMemberPostConditions(inputSlot, property.GetMethod);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:459]: I'm confused about this bit in light of the `gotoNodeWithCurrentStae` after it. It seems like when you apply the member postconditions, you can get into a split state, but then we will `gotoNodeWithCurrentState`, which expects a non-split state. What am I missing?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 20]
```diff
@@ -24605,6 +24605,1467 @@ void Init()
...
+        [Fact, WorkItem(51020, "https://github.com/dotnet/roslyn/issues/51020")]
+        public void PatternOnTuple_IsExpression_LongTuple()
+        {
+            var src = @"
+#nullable enable
+public class C
+{
+    void M(object? o, object o2)
+    {
+        if ((o2, o2, o2, o2, o2, o2, o2, o) is (null, null, null, null, null, null, null, not null))
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:24980]: what happens when contradictory tests are performed on the same "input"? i.e.

```cs
    public void M(object? obj) {
        if ((obj, obj) is (null, not null)) { obj.ToString(); }
    }
```

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 21]
```diff
@@ -24605,6 +24605,1467 @@ void Init()
...
+public class Container
+{
+    public C? Item;
+
+    public string M()
+    {
+        switch (this)
+        {
+            case { Item: { IsOk: true } item }:
+                return item.Value;
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:24694]: should we have a version of this test which returns `this.Item.Value` here?

**Comment by RikkiGibson** [comment]
Done with review pass, just had some minor questions and suggestions.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs (hunk 16 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:252]: I agree this is tempting. I'm not sure why AbstractFlowPass doesn't do something like that. 
But given the existing pattern, I prefer to stay with separate fields. It's more consistent and possibly clearer. This is something I hope we could make easier with discriminated union types.

---
In reply to: [573999711](https://github.com/dotnet/roslyn/pull/51001#discussion_r573999711) [](ancestors = 573999711)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs (hunk 17 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:405]: Right, we want to compute the input slots once for the whole DAG

---
In reply to: [574006304](https://github.com/dotnet/roslyn/pull/51001#discussion_r574006304) [](ancestors = 574006304)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs (hunk 19 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:459]: `gotoNodeWithCurrentState` can handle split states, but it currently asserts out if we have a split state *and* we're merging two branches (the next node already has some accumulated state attached to it). I've tried a bunch of things but could not hit that combination.

---
In reply to: [574024479](https://github.com/dotnet/roslyn/pull/51001#discussion_r574024479) [](ancestors = 574024479)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 20 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:24980]: Haha, that's creative :-) I'll add the scenario. 
The last test (in our unspecified evaluation order) wins.

---
In reply to: [574028591](https://github.com/dotnet/roslyn/pull/51001#discussion_r574028591) [](ancestors = 574028591)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs (hunk 18 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:510]: Without this check, only a couple of tests break.
Here's one:
```
    int M1(string s1, string s2)
    {
        return (s1, s2) switch {
            (string x, string y) => x.Length + y.Length
            };
    }
```
That scenario gets a new warning:
```
 // (6,25): warning CS8655: The switch expression does not handle some null inputs (it is not exhaustive). For example, the pattern '(null, _)' is not covered.
                    //         return (s1, s2) switch {
                    Diagnostic(ErrorCode.WRN_SwitchExpressionNotExhaustiveForNull, "switch").WithArguments("(null, _)").WithLocation(6, 25),
```
So this isn't so much "treating pure null tests seriously" as "consider branches unreachable when there is an implicit null-test but the input is not-null".

Pure null tests are handled in the `BoundDagExplicitNullTest` arm below. Those will change an input from not-null to maybe-null.

---
In reply to: [574012059](https://github.com/dotnet/roslyn/pull/51001#discussion_r574012059) [](ancestors = 574012059)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 21 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:24694]: Added test above (`MemberNotNullWhenTrue_RecursivePattern_Nested`)


---
In reply to: [574034771](https://github.com/dotnet/roslyn/pull/51001#discussion_r574034771) [](ancestors = 574034771)

**Review by RikkiGibson — APPROVED** [review]
